llvm/docs/LangRef.rst

```diff
@@ -14531,6 +14531,63 @@ The arguments (``%a`` and ``%b``) may be of any integer type or a vector with
 integer element type. The argument types must match each other, and the return
 type must match the argument type.
 
+.. _int_sthreecmp:
+
+'``llvm.sthreecmp.*``' Intrinsic
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Syntax:
+"""""""
+
+This is an overloaded intrinsic. You can use ``@llvm.sthreecmp`` on any
+integer bit width or any vector of integer elements.
+
+::
+
+      declare i32 @llvm.sthreecmp.i32(i32 %a, i32 %b)
+      declare <4 x i32> @llvm.sthreecmp.v4i32(<4 x i32> %a, <4 x i32> %b)
+
+Overview:
+"""""""""
+
+Return ``-1`` if ``%a`` is less than ``%b``, ``0`` if they are equal, and 
+``1`` if ``%a`` is greater than ``%b``. Vector intrinsics operate on a per-element basis. 
+
+Arguments:
+""""""""""
+
+The arguments (``%a`` and ``%b``) may be of any signed integer type or a vector with
+integer element type. The argument types must match each other, and the return
+type must be at least as wide as ``i2``, to uphold the ``-1`` return value.
+
+.. _int_uthreecmp:
+
+'``llvm.uthreecmp.*``' Intrinsic
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Syntax:
+"""""""
+
+This is an overloaded intrinsic. You can use ``@llvm.sthreecmp`` on any
+integer bit width or any vector of integer elements.
+
+::
+
+      declare i2 @llvm.uthreecmp.i32(i32 %a, i32 %b)
```

Comment:
Is this meant to be `i32`? I think there is also merit in deciding whether to use `i32` to represent the result or a type like `i2` because that would be the minimum size required. I feel like `i8` is probably the best type to use for the return type.

Reply:
Oh, thanks for pointing that out. Nikita previously mentioned it should be defined as
```
; Returns 1 if (%a ugt %b), -1 if (%a ult %b), 0 if (%a eq %b)
iN @llvm.ucmp.iN.iM(iM %a, iM %b)

; Returns 1 if (%a sgt %b), -1 if (%a slt %b), 0 if (%a eq %b)
iN @llvm.scmp.iN.iM(iM %a, iM %b)
```
but I'd love to rescue some of the parametricity on the result type `iN` to leave more SIMD optimizations on the table in the future.

Reply:
I forwarded this question to the Discourse thread: https://discourse.llvm.org/t/rfc-add-3-way-comparison-intrinsics/76685/7?u=nikic

I assumed we're going to use two type overloads for the result and arg type. I hadn't really considered the possibility of using a fixed type.

Reply:
Sweet. I'll look into doing some tutorials for analysis with GlobalISel while people mull it over. Thanks!

Reply:
There wasn't much feedback regarding the return type on the discourse thread, but it seems like there is a mild preference to have the result type overload over a fixed type.

The way to implement would be like this in Intrinsics.td:
```
def int_scmp : DefaultAttrsIntrinsic<
    [llvm_anyint_ty], [llvm_anyint_ty, LLVMMatchType<1>],
    [IntrNoMem, IntrSpeculatable, IntrWillReturn]>;
```
and then add a check in Verifier.cpp that a) the return type and the first argument type have the same number of elements (if they are vectors) and b) the return scalar type has width at least 2.

Reply:
Sweet, I'll try that.

Reply:
Double checking: That check should be a
```cpp
case Intrinsic::scmp:
case Intrinsic::ucmp:
```
under the [Verifier::visitIntrinsicCall(Intrinsic::IK ID, CallBase &Call)](https://github.com/llvm/llvm-project/blob/193b3d6733b7bf606c70749b1b65b6a0daae97d5/llvm/lib/IR/Verifier.cpp#L5150) function, correct?

Reply:
Yeah.

Reply:
Alright, put in a good first effort.